include/ginkgo/core/config/data.hpp

```diff
@@ -0,0 +1,268 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_DATA_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_DATA_HPP_
+
+
+#include <exception>
+#include <string>
+#include <type_traits>
+#include <typeinfo>
+
+
+namespace gko {
+namespace config {
+
+
+/** @file data.hpp
+ * It implements the base type for property tree. It only handles std::string,
+ * double, long long int, bool, monostate(empty). Because it can be handled by
+ * std::variant in C++17 directly, this file tries to use the same function
+ * signature such that we can replace data without breaking public interface
+ * when we decide to use C++17.
+ */
```

Comment:
Do we need to expose `data`, or can we hide it behind `property_tree`'s interface? If we put it into `namespace detail`, we don't need to worry about C++17 compatibility

Reply:
In general, I do not think we need to consider the C++17 compatibility because it is a future requirement and we do not have a certain time for it. I know @MarcelKoch would like to push C++17, so I try to have it without breaking the public interface in C++17.

Reply:
I would say it's possible we will move to C++17 before Ginkgo 2.0, so we should plan for that possibility